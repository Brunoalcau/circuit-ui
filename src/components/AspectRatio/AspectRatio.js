import React, { Children } from 'react';
import PropTypes from 'prop-types';
import styled, { ClassNames, css } from 'react-emotion';

import { childrenPropType } from '../../util/shared-prop-types';

console.log(ClassNames);

const wrapperBaseStyles = () => css`
  display: block;
  position: relative;
  overflow: hidden;
  height: auto;
  width: 100%;
`;

const wrapperAspectRatioStyles = ({ aspectRatio }) =>
  aspectRatio &&
  css`
    height: 0;
    width: 100%;
    padding-top: ${Math.round((1 / aspectRatio) * 100)}%;
  `;

const Wrapper = styled('div')(wrapperBaseStyles, wrapperAspectRatioStyles);

const childBaseStyles = cssClassName => cssClassName`
  display: block;
  height: auto;
  max-height: 100%;
  width: 100%;
`;

const childAspectRatioStyles = (cssClassName, { aspectRatio }) =>
  aspectRatio &&
  cssClassName`
    position: absolute;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
    width: 100%;
    height: 100%;
    max-height: none;
    object-fit: cover;
    z-index: 2;
  `;

function AspectRatio({ aspectRatio, children, className }) {
  if (!children) {
    return null;
  }

  const [child, ...restChildren] = Children.toArray(children);
  return (
    <Wrapper aspectRatio={aspectRatio} className={className}>
      <ClassNames>
        {({ css: cssClassName, cx }) =>
          React.cloneElement(child, {
            className: cx(
              childBaseStyles(cssClassName),
              childAspectRatioStyles(cssClassName, { aspectRatio })
            )
          })
        }
      </ClassNames>
      {restChildren}
    </Wrapper>
  );
}

AspectRatio.propTypes = {
  children: childrenPropType,
  aspectRatio: PropTypes.number,
  className: PropTypes.string
};

export default AspectRatio;
