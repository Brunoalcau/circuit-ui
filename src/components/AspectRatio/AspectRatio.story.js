import React from 'react';
import { storiesOf } from '@storybook/react';
import { number } from '@storybook/addon-knobs/react';
import styled from 'react-emotion';

import { GROUPS } from '../../../.storybook/hierarchySeparators';

import AspectRatio from './AspectRatio';

const Background = styled('div')`
  background: lightgrey;
`;

storiesOf(`${GROUPS.COMPONENTS}|AspectRatio`, module).add('AspectRatio', () => (
  <div style={{ width: '50vw' }}>
    <AspectRatio aspectRatio={number('Aspect ratio', 1.5)}>
      <Background />
    </AspectRatio>
  </div>
));
