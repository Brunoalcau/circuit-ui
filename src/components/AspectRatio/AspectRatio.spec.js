import React from 'react';

import AspectRatio from './AspectRatio';

describe.skip('AspectRatio', () => {
  it('should render with default styles', () => {
    const actual = create(
      <AspectRatio>
        <div />
      </AspectRatio>
    );
    expect(actual).toMatchSnapshot();
  });

  it('should render with fixed aspect ratio styles', () => {
    const actual = create(
      <AspectRatio aspectRatio={1.618}>
        <div />
      </AspectRatio>
    );
    expect(actual).toMatchSnapshot();
  });

  it('should not render without children', () => {
    const actual = create(<AspectRatio spectRatio={1.618} />);
    expect(actual).toMatchSnapshot();
  });
});
