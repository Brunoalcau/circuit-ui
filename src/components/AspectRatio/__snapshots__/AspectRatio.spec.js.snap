// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AspectRatio should not render without children 1`] = `null`;
